### PR TITLE
Implement percentage-closer filtering (PCF) for point lights.

### LINKED ISSUE #3628: Percentage Closer Filtering for shadows
# Description

- bevy's shadows currently use 'hardware' 2x2 Percentage Closer Filtering resulting in anti-aliased edges of shadows. However, the filtering is insufficient for smooth shadow edges and more extensive/advanced filtering patterns can be used to producer smoother-looking shadows.
  - 'hardware' is in single quotes because apparently in modern GPUs there may not be dedicated silicon for this purpose, but rather the functionality is software in drivers that leverages more generic hardware acceleration functionality.

## Background and Reference Material

- **Percentage Closer Filtering** involves taking multiple samples from the shadow map and comparing with an interpolated mesh surface depth projected into the frame of reference of the light and calculating the percentage of samples in the depth buffer that are closer to the light (i.e. casting shadow) than the mesh surface.
- Sampling from multiple shadow map texels away from the one that 'covers' the fragment on the mesh surface and using the same 'biased' (offset) fragment position easily produces shadow acne/self-shadowing even if some bias configuration works well for sampling texels close to the centre of the kernel.
- Roughly-speaking, more samples are needed to gain more information in order to produce a smoother result. Sampling the same texels repeatedly doesn't help, so more samples roughly means sampling with a larger radius.
- _Percentage Closer Soft Shadows_ (a separate issue) scale both the occluder search radius and the filtering radius based on the distance to the light, and occluder, and the size of the light itself. This can result in large radius sampling kernels which easily produces artifacts due to self-shadowing
- Regular grid sampling patterns produce filtering artifacts like banding.
- Irregular sampling patterns such as blue noise or poisson disc require fewer samples for the same visual quality, don't produce the same banding artifacts, and performance can be scaled nicely (you can choose an arbitrary number of samples to take).
- The Witness has a really nice and smooth PCF sampling and post-filtering (as in, filtering the results of sampling) algorithm (http://the-witness.net/news/2013/09/shadow-mapping-summary-part-1/) that is among the best I have seen for surfaces without much texture (combining textured surfaces and noisy shadows is less visually jarring than noisy shadows on flat shaded surfaces where the noise stands out). I think this would be a good default. It requires relatively few samples to look good too.
- Randomly-rotated (using blue noise for the rotations) poisson disc / direct blue noise sampling would work well with Temporal Anti-Aliasing and would also require few samples, but is visually noisy without it.
- Cubemaps are left-handed!!!
- Biasing
  - **Receiver Plane Depth Bias** (Isidoro 2006, Page 36) - This can be used for 2D shadow maps and it is quite simple to implement and compute: https://developer.amd.com/wordpress/media/2012/10/Isidoro-ShadowMapping.pdf However, I do not know if it is possible to adapt this to be used for cube shadow maps for point lights so I went in search of alternatives.
  - **Adaptive Bias for Shadow Maps** (Dou et al 2014) - https://jcgt.org/published/0003/04/08/ This approach is very intuitive and works well. It assumes the mesh surface is locally flat, calculates the gradient of the surface to calculate an 'optimal' fragment depth at the surface at the offset position in light view x/y, and then uses an additional 'adaptive epsilon' that is based on the gradient of the depth projection being used to account for, if I recall correctly, the precision distribution in order to move the fragment position just closer to the light to avoid self-shadowing.
  - **Adaptive Bias for Soft Shadows** (Ehm et al. 2015) - https://w3-o.cs.hm.edu/users/nischwit/public_html/AdaptiveDepthBias_WSCG.pdf - included here as it details a good further enhancement to _Adaptive Bias for Shadow Maps_ though its extension to _Percentage Closer Soft Shadows_ is not relevant here.

## Solution

- https://github.com/superdump/bevy/tree/shadow-pcf
  - A large bunch of experiments to try to get both point and directional light shadow filtering to work well with large filter kernels
  - Implements left-handed cube maps. This was a long-and-hard-fought realisation as documentation for cubemaps is awful. They are left-handed and so we try to force them to work with right-handed projections, rather accept they are left-handed and go with it. It makes for much more comprehensible code that one can reason about!
  - Implements the above-mentioned techniques (receiver plane depth bias, adaptive bias for shadow maps, part of adaptive bias for soft shadows) and some regular-grid and blue noise sampling pattern techniques
- https://github.com/superdump/bevy/tree/poisson-pcf
  - Implements the blue noise random rotated poisson disc and _The Witness_ sampling techniques

## Next Steps

- Clean up the conversion of cube map handling to be left-handed and make a PR
- Clean up _Adaptive Depth Bias for Shadow Maps_ with the improvement to scale the adaptive epsilon by `min(1 / dot(light dir, frag normal)^2, 100)` as introduced in _Adaptive Depth Bias for Soft Shadows_, make this depth biasing an option, benchmark it, and make a PR
- Clean up the _The Witness_ PCF sampling approach, and make a PR
- Investigate the value of direct blue noise sampling and blue noise random rotated poisson disc sampling and decide whether they are worth integrating now or should wait for reevaluation when we have _Temporal Anti-Aliasing_

### PR DESCRIPTION
**Pull request by pcwalton**
I ported the two existing PCF techniques to the cubemap domain as best I could. Generally, the technique is to create a 2D orthonormal basis using Gram-Schmidt normalization, then apply the technique over that basis. The results look fine, though the shadow bias often needs adjusting.

For comparison, Unity uses a 4-tap pattern for PCF on point lights of (1, 1, 1), (-1, -1, 1), (-1, 1, -1), (1, -1, -1). I tried this but didn't like the look, so I went with the design above, which ports the 2D techniques to the 3D domain. There's surprisingly little material on point light PCF.

I've gone through every example using point lights and verified that the shadow maps look fine, adjusting biases as necessary.

Fixes #3628.

---

### Added
* Shadows from point lights now support percentage-closer filtering (PCF), and as a result look less aliased.

### Changed
* `ShadowFilteringMethod::Castano13` and `ShadowFilteringMethod::Jimenez14` have been renamed to `ShadowFilteringMethod::Gaussian` and `ShadowFilteringMethod::Temporal` respectively.

## Migration Guide

* `ShadowFilteringMethod::Castano13` and `ShadowFilteringMethod::Jimenez14` have been renamed to `ShadowFilteringMethod::Gaussian` and `ShadowFilteringMethod::Temporal` respectively.

### CODE REVIEW

**Comment by pcwalton** [comment]
Before:

Screenshot 2024-04-08 120914

After:

Screenshot 2024-04-08 120819

**Comment by superdump** [comment]
Please don’t merge this until I have reviewed it. It’s something I particularly care about.

**Comment by pcwalton** [comment]
I looked at three.js. It uses a 3x3 box filter: https://github.com/mrdoob/three.js/blob/6421f3a49ebed3d07de781776f47624e6d26e896/src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl.js#L295-L309

I also looked at Blender. It doesn't seem to have any PCF for point lights at all.

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl [hunk 1]
```diff
@@ -26,6 +26,40 @@ fn sample_shadow_map_hardware(light_local: vec2<f32>, depth: f32, array_index: i
...
+    vec2(-0.375, -0.625),
+    vec2(-0.625,  0.625),
+    vec2(-0.875, -0.125),
+    vec2( 0.375,  0.875),
+    vec2( 0.875, -0.875),
+);
+
+// And these are the coefficients corresponding to the probability distribution
+// function of a 2D Gaussian lobe with zero mean and the identity covariance matrix
+// ] at those points.
```
Comment (superdump) [inline crates/bevy_pbr/src/render/shadow_sampling.wgsl:51]:
```suggestion
// at those points.
```

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl [hunk 2]
```diff
@@ -126,3 +167,166 @@ fn sample_shadow_map(light_local: vec2<f32>, depth: f32, array_index: i32, texel
...
+    return sample_shadow_cubemap_hardware(
+        light_local + position.x * x_basis + position.y * y_basis,
+        depth,
+        light_id
+    ) * coeff;
+}
+
+// This more or less does what Castano13 does, but in 3D space. Castano13 is
+// essentially an optimized 2D Gaussian filter that takes advantage of the
+// bilinear filtering hardware to reduce the number of taps needed. This trick
```
Comment (superdump) [inline crates/bevy_pbr/src/render/shadow_sampling.wgsl:202]: I’d say the filter has the same number of taps but the texture samples account for multiple as they are bilinearly filtered. I feel ok with considering samples to implicitly mean texture samples.
```suggestion
// bilinear filtering hardware to reduce the number of samples needed. This trick
```

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl [hunk 3]
```diff
@@ -126,3 +167,166 @@ fn sample_shadow_map(light_local: vec2<f32>, depth: f32, array_index: i32, texel
...
+// points in the spiral pattern after first creating a 2D orthonormal basis
+// along the principal light direction.
+fn sample_shadow_cubemap_temporal(
+    light_local: vec3<f32>,
+    depth: f32,
+    texel_size: f32,
+    distance_to_light: f32,
+    light_id: u32,
+) -> f32 {
+    // Create an orthonormal basis via Gram-Schmidt normalization so we can
```
Comment (superdump) [inline crates/bevy_pbr/src/render/shadow_sampling.wgsl:261]: Maybe put this in a reusable function given a normal and up vector.

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl [hunk 4]
```diff
@@ -126,3 +167,166 @@ fn sample_shadow_map(light_local: vec2<f32>, depth: f32, array_index: i32, texel
...
+// The shadow maps have no mipmaps so Level just samples from LOD 0.
+fn sample_shadow_cubemap_hardware(light_local: vec3<f32>, depth: f32, light_id: u32) -> f32 {
+#ifdef NO_CUBE_ARRAY_TEXTURES_SUPPORT
+    return textureSampleCompare(view_bindings::point_shadow_textures, view_bindings::point_shadow_textures_sampler, light_local, depth);
+#else
+    return textureSampleCompareLevel(view_bindings::point_shadow_textures, view_bindings::point_shadow_textures_sampler, light_local, i32(light_id), depth);
+#endif
+}
+
+fn tap_shadow_cubemap(
```
Comment (superdump) [inline crates/bevy_pbr/src/render/shadow_sampling.wgsl:184]:
```suggestion
fn sample_shadow_cubemap_at_offset(
```

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl [hunk 5]
```diff
@@ -126,3 +167,166 @@ fn sample_shadow_map(light_local: vec2<f32>, depth: f32, array_index: i32, texel
...
+// doesn't apply to cubemaps, so we manually apply a Gaussian filter over the
+// standard 8xMSAA pattern instead.
+fn sample_shadow_cubemap_gaussian(
+    light_local: vec3<f32>,
+    depth: f32,
+    texel_size: f32,
+    distance_to_light: f32,
+    light_id: u32,
+) -> f32 {
+    // Create an orthonormal basis via Gram-Schmidt normalization so we can
```
Comment (superdump) [inline crates/bevy_pbr/src/render/shadow_sampling.wgsl:212]: This isn’t Gram-Schmidt I think. That involves projection and subtraction if I remember correctly.

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl [hunk 6]
```diff
@@ -26,6 +26,40 @@ fn sample_shadow_map_hardware(light_local: vec2<f32>, depth: f32, array_index: i
 #endif
 }
 
+// Numbers determined by trial and error that gave nice results.
+const SPOT_SHADOW_TEXEL_SIZE: f32 = 0.0134277345;
```
Comment (superdump) [inline crates/bevy_pbr/src/render/shadow_sampling.wgsl:30]: These aren’t texel sizes, they’re fudge factors right?

File: examples/3d/3d_shapes.rs [hunk 7]
```diff
@@ -70,6 +70,7 @@ fn setup(
             shadows_enabled: true,
             intensity: 10_000_000.,
             range: 100.0,
+            shadow_depth_bias: 0.4,
```
Comment (superdump) [inline examples/3d/3d_shapes.rs:73]: I don’t remember the default but we could change it if needed.

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl (hunk 5 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/render/shadow_sampling.wgsl:212]: It's equivalent to Gram-Schmidt. https://math.stackexchange.com/a/1849294

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl (hunk 6 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/render/shadow_sampling.wgsl:30]: In the case of spot shadows, which I didn't change, it was passed into `sample_shadow_map` as `texel_size`. In other invocations of `sample_shadow_map`, the actual texel size is passed into that parameter. So either it actually is the texel size or something is misnamed. What do you want me to do here?

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl (hunk 5 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/render/shadow_sampling.wgsl]: I removed the comment.

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl (hunk 6 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/render/shadow_sampling.wgsl:31]: In the case of point light PCF I renamed it to `SCALE`.

File: examples/3d/3d_shapes.rs (hunk 7 above)
Comment (pcwalton) [reply examples/3d/3d_shapes.rs]: It was 0.02. I changed the default to 0.08, which was the lowest value that made the shadow acne go away in those demos (I think anyway; it's hard to tell).

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl (hunk 3 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/render/shadow_sampling.wgsl]: It's been moved to a new function `orthonormalize`.

File: crates/bevy_pbr/src/light/point_light.rs [hunk 8]
```diff
@@ -51,6 +51,6 @@ impl Default for PointLight {
 }
 
 impl PointLight {
-    pub const DEFAULT_SHADOW_DEPTH_BIAS: f32 = 0.02;
+    pub const DEFAULT_SHADOW_DEPTH_BIAS: f32 = 0.08;
```
Comment (superdump) [inline crates/bevy_pbr/src/light/point_light.rs:54]: Does this produce any Peter-panning?

File: crates/bevy_render/src/maths.wgsl [hunk 9]
```diff
@@ -27,3 +27,16 @@ fn mat2x4_f32_to_mat3x3_unpack(
...
+
+// Creates an orthonormal basis given a Z vector and an up vector (which becomes
+// Y after orthonormalization).
+//
+// The results are equivalent to the Gram-Schmidt process [1].
+//
+// [1]: https://math.stackexchange.com/a/1849294
+fn orthonormalize(z_unnormalized: vec3<f32>, up: vec3<f32>) -> mat3x3<f32> {
+    let z_basis = normalize(z_unnormalized);
+    let x_basis = normalize(cross(z_basis, up));
```
Comment (superdump) [inline crates/bevy_render/src/maths.wgsl:39]: I’m trying to think of this necessarily works for both left- and right-handed coordinate systems. I think it does…

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl (hunk 6 above)
Comment (superdump) [reply crates/bevy_pbr/src/render/shadow_sampling.wgsl:31]: Ah, I see it now. Hmm, leave it as it is for now. We can fix it separately.

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl [hunk 10]
```diff
@@ -126,3 +168,160 @@ fn sample_shadow_map(light_local: vec2<f32>, depth: f32, array_index: i32, texel
...
+    var up = vec3(0.0, 1.0, 0.0);
+    if (dot(up, normalize(light_local)) > 0.99) {
+        up = vec3(1.0, 0.0, 0.0);   // Avoid creating a degenerate basis.
+    }
+    let basis = orthonormalize(light_local, up) * scale * distance_to_light;
+
+    var sum: f32 = 0.0;
+    sum += sample_shadow_cubemap_at_offset(
+        D3D_SAMPLE_POINT_POSITIONS[0], D3D_SAMPLE_POINT_COEFFS[0],
+        basis[0], basis[1], light_local, depth, light_id);
```
Comment (superdump) [inline crates/bevy_pbr/src/render/shadow_sampling.wgsl:224]: I hope that basis[2] is never used means its calculation is optimised away. :)

File: crates/bevy_pbr/src/light/point_light.rs (hunk 8 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/light/point_light.rs:54]: Technically no, from what I've seen, but there is a little bit of a gap in `3d_scene`:

Screenshot 2024-04-09 100610

Let me know what you'd like me to do: (1) adjust the default bias down and add the bias back in `3d_shapes`, or (2) leave as is.

File: crates/bevy_render/src/maths.wgsl (hunk 9 above)
Comment (pcwalton) [reply crates/bevy_render/src/maths.wgsl:39]: It will, in that it will produce *some* orthonormal basis. It might not be the orthonormal basis that you want it to, though, if you have some extra requirements. In our case it works because we don't really care about what the orthonormal basis is as long as the Z vector is retained.

File: crates/bevy_pbr/src/render/shadow_sampling.wgsl (hunk 10 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/render/shadow_sampling.wgsl:224]: Well, you can't optimize away its calculation, because it's used for the construction of the other two columns. As written, this is a dead store. However, this function will certainly be inlined, and the matrix will all be in GPU registers. So the register allocation liveness range analysis will trivially detect that it isn't used and returning the extra value will have no effect.

tl;dr this has no performance impact.

File: crates/bevy_pbr/src/light/point_light.rs (hunk 8 above)
Comment (superdump) [reply crates/bevy_pbr/src/light/point_light.rs:54]: That looks good enough to me.

**Review by superdump — APPROVED** [review]

**Review by IceSentry — APPROVED** [review]
Tested on windows 10, webgl and webgpu. Everything looks good. Code quality is great. I like changing the names to Gaussian and Temporal. I don't understand the math enough to comment on it though.

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1298 if you'd like to help out.
